lib/property_table/persist.ex

```diff
@@ -0,0 +1,229 @@
+    encoded = PersistFile.encode_binary(binary_data)
+
+    File.write!(stable_path, encoded, [:binary, :sync])
+
+    :ok
+  rescue
+    e ->
+      Logger.error("Failed to persist table to disk: #{e}")
+      :error
+  end
+
+  @spec restore_from_disk(reference() | atom(), Keyword.t()) :: :ok | {:error, atom()}
+  def restore_from_disk(table, options) do
+    options = take_options(options)
+
+    stable_path = get_path(:stable, options)
+    backup_path = get_path(:backup, options)
+
+    # Reduce over all the possible paths we want to try and restore from
+    attempts = [stable_path, backup_path]
+
+    result =
+      Enum.reduce_while(attempts, {:error, :failed_to_load}, fn path, _ ->
+        if File.exists?(path) do
+          try_restore_tabfile(path)
+        else
+          Logger.warn("File #{path} does not exist! Trying a backup file...")
+          {:cont, {:error, :enoent}}
+        end
+      end)
+
+    case result do
+      {:ok, data} ->
+        ^table = :ets.new(table, [:named_table, :public])
+
+        # Insert the restored properties at the current timestamp
+        timestamp = System.monotonic_time()
+
+        Enum.each(data, fn {property, value} ->
+          :ets.insert(table, {property, value, timestamp})
+        end)
+
+        :ok
+
+      error ->
+        error
+    end
+  rescue
+    e ->
+      Logger.error("Failed to restore table from disk: #{e}")
+      {:error, :failed_to_restore}
+  end
+
+  @spec save_snapshot(reference() | atom(), Keyword.t()) ::
+          {:ok, String.t()} | :error | no_return()
+  def save_snapshot(table, options) do
+    options = take_options(options)
+    persist_to_disk(table, options)
+
+    snapshot_id = :crypto.strong_rand_bytes(8) |> Base.encode16()
+
+    full_snapshot_name = "#{snapshot_id}"
+    stable_path = get_path(:stable, options)
+    snapshot_path = get_path(:snapshot, options, full_snapshot_name)
+
+    # Move file to snapshot directory
+    Logger.debug("Creating table file snapshot: #{snapshot_path}")
+    File.copy!(stable_path, snapshot_path)
+
+    maybe_clean_old_snapshots(options)
+
+    {:ok, snapshot_id}
+  rescue
+    e ->
+      Logger.error("Failed to save snapshot to disk: #{e}")
+      :error
+  end
+
+  @spec restore_snapshot(Keyword.t(), String.t()) :: :ok | {:error, :enoent} | no_return()
+  def restore_snapshot(options, snapshot_id) do
+    options = take_options(options)
+    stable_path = get_path(:stable, options)
+    snapshots_path = get_path(:snapshot, options) |> Path.join("#{snapshot_id}")
+    found_snapshot = Path.wildcard(snapshots_path)
+
+    if length(found_snapshot) != 1 do
+      {:error, :enoent}
+    else
+      [to_restore] = found_snapshot
+
+      # Copy the found snapshot in place of the current stable db path
+      Logger.debug("Restoring table file snapshot: #{to_restore} => #{stable_path}")
+      File.copy!(to_restore, stable_path)
+
+      :ok
+    end
```

Comment:
It seems like `File.copy/2` should already do the check for whether the snapshot exists, so the wildcard check isn't necessary.

```suggestion
    snapshots_path = get_path(:snapshot, options) |> Path.join(snapshot_id)

    # Copy the found snapshot in place of the current stable db path
    Logger.debug("Restoring table file snapshot: #{to_restore} => #{stable_path}")
    case File.copy(snapshots_path, stable_path) do
      {:ok, _} -> :ok
      error -> error
    }
```

Having said this, I'm wondering whether the snapshot should be loaded first just in case it's corrupt. Then, if it's not corrupt, the file is copied over.

Reply:
Good point about validating the file before we move it into place, let me add that in...